server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/model/MockSMTPBehaviorInformation.java

```diff
@@ -0,0 +1,165 @@
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.mock.smtp.server.model;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import com.google.common.base.Preconditions;
+
+public class MockSMTPBehaviorInformation {
+
+    interface RemainingAnswersCounter {
+
+        void decrease();
+
+        boolean hasRemainingAnswers();
+
+        Optional<Integer> getValue();
+    }
+
+    static class UnlimitedAnswersCounter implements RemainingAnswersCounter {
+
+        @Override
+        public void decrease() {
+            // do nothing
+        }
+
+        @Override
+        public boolean hasRemainingAnswers() {
+            return true;
+        }
+
+        @Override
+        public Optional<Integer> getValue() {
+            return Optional.empty();
+        }
+
+        @Override
+        public final boolean equals(Object o) {
+            if (o instanceof UnlimitedAnswersCounter) {
+                return true;
+            }
+            return false;
+        }
+
+        @Override
+        public final int hashCode() {
+            return Objects.hash(UnlimitedAnswersCounter.class);
+        }
+    }
+
+    static class LimitedAnswersCounter implements RemainingAnswersCounter {
+
+        private final AtomicInteger remainingCount;
+
+        LimitedAnswersCounter(int remainingCount) {
+            Preconditions.checkArgument(remainingCount > 0, "remainingCount should be positive");
+
+            this.remainingCount = new AtomicInteger(remainingCount);
+        }
+
+        @Override
+        public void decrease() {
+            remainingCount.updateAndGet(currentCount -> {
+                if (currentCount > 0) {
+                    return currentCount - 1;
+                }
+                throw new IllegalStateException("remainingCount is already being zero, you can not decrease more");
+            });
+        }
+
+        @Override
+        public boolean hasRemainingAnswers() {
+            return remainingCount.get() > 0;
+        }
+
+        @Override
+        public Optional<Integer> getValue() {
+            return Optional.of(remainingCount.get());
+        }
+
+        @Override
+        public final boolean equals(Object o) {
```

Comment:
It's not because 2 counter happens to have the same value that they happen to be the same.

They are not POJO.

Please don't implement equals/hashCode in them.